protocol/bc/txinput.go

```diff
@@ -18,279 +18,233 @@ type (
 
 	TypedInput interface {
 		IsIssuance() bool
+		readWitness(r io.Reader, assetVersion uint64) error
 	}
 )
 
 var errBadAssetID = errors.New("asset ID does not match other issuance parameters")
 
-func (t *TxInput) AssetAmount() AssetAmount {
-	if ii, ok := t.TypedInput.(*IssuanceInput); ok {
-		return AssetAmount{
-			AssetID: ii.AssetID(),
-			Amount:  ii.Amount,
-		}
+func (t *TxInput) writeTo(w io.Writer, serflags uint8) error {
+	_, err := blockchain.WriteVarint63(w, t.AssetVersion)
+	if err != nil {
+		return err
 	}
-	si := t.TypedInput.(*SpendInput)
-	return si.AssetAmount
-}
-
-func (t *TxInput) AssetID() AssetID {
-	if ii, ok := t.TypedInput.(*IssuanceInput); ok {
-		return ii.AssetID()
+	_, err = blockchain.WriteExtensibleString(w, func(w io.Writer) error {
+		return t.WriteInputCommitment(w)
+	})
+	if err != nil {
+		return err
 	}
-	si := t.TypedInput.(*SpendInput)
-	return si.AssetID
-}
-
-func (t *TxInput) Amount() uint64 {
-	if ii, ok := t.TypedInput.(*IssuanceInput); ok {
-		return ii.Amount
+	_, err = blockchain.WriteVarstr31(w, t.ReferenceData)
+	if err != nil {
+		return err
 	}
-	si := t.TypedInput.(*SpendInput)
-	return si.Amount
-}
-
-func (t *TxInput) ControlProgram() []byte {
-	if si, ok := t.TypedInput.(*SpendInput); ok {
-		return si.ControlProgram
+	if serflags&SerWitness != 0 {
+		_, err = blockchain.WriteExtensibleString(w, func(w io.Writer) error {
+			return t.writeInputWitness(w)
```

Comment:
can `t.WriteInputWitness` be used as the param to `WriteExtensibleString` directly?